pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,194 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"knative.dev/pkg/logging"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	podName             = "name"
+	podNameSpace        = "namespace"
+	ownerSelfLink       = "owner"
+	podHostName         = "node"
+	podProvisioner      = "provisioner"
+	podHostZone         = "zone"
+	podHostArchitecture = "arch"
+	podHostCapacityType = "capacitytype"
+	podHostInstanceType = "instancetype"
+	podPhase            = "phase"
+	podLabels           = "podLabels"
+)
+
+var (
+	podGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "pods",
+			Name:      "podstate",
+			Help:      "Pod state.",
+		},
+		getLabelNames(),
+	)
+)
+
+// Controller for the resource
+type Controller struct {
+	KubeClient   client.Client
+	CoreV1Client corev1.CoreV1Interface
+	LabelsMap    map[types.NamespacedName]*prometheus.Labels
+}
+
+func init() {
+	crmetrics.Registry.MustRegister(podGaugeVec)
+}
+
+func getLabelNames() []string {
+	return []string{
+		podName,
+		podNameSpace,
+		ownerSelfLink,
+		podHostName,
+		podProvisioner,
+		podHostZone,
+		podHostArchitecture,
+		podHostCapacityType,
+		podHostInstanceType,
+		podPhase,
+		podLabels,
+	}
+
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, coreV1Client corev1.CoreV1Interface) *Controller {
+	newcontroller := Controller{
+		KubeClient:   kubeClient,
+		CoreV1Client: coreV1Client,
+		LabelsMap:    make(map[types.NamespacedName]*prometheus.Labels),
+	}
+	return &newcontroller
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("podmetrics").With("pod", req.Name))
+
+	// Retrieve pod from reconcile request
+	pod := &v1.Pod{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to pod deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				podGaugeVec.Delete(*labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after pod labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
```

Comment:
I'm not grokking this. Can't we just update it?

Reply:
Not really. Prometheus treats metric with different label values as different active time series. So if we have new label values (representing the state of the pod), it will create a new gauge in gaugeVec. We need to delete the old one so that Prometheus will not create wrong counts.